cmds/core/sort/sort.go

```diff
@@ -65,45 +97,84 @@ func readInput(w io.Writer, f *os.File, args ...string) error {
 			fileContents = append(fileContents, "\n")
 		}
 	}
-	if err := writeOutput(w, sortAlgorithm(strings.Join(fileContents, ""))); err != nil {
+
+	s := strings.Join(fileContents, "")
+	if len(s) > 0 && s[len(s)-1] == '\n' {
+		s = s[:len(s)-1] // remove newline terminator
+	}
+
+	if c.params.ordered {
+		lines := strings.Split(s, "\n")
+		if !sort.IsSorted(sort.StringSlice(lines)) {
+			return errNotOrdered
+		}
+
+		if c.params.unique && len(lines) > 1 {
+			for i := 1; i < len(lines); i++ {
+				if lines[i] == lines[i-1] {
+					return errNotOrdered
+				}
+			}
+		}
+
+		return nil
+	}
+
+	if err := c.writeOutput(c.stdout, c.sortAlgorithm(s)); err != nil {
 		return err
 	}
 	return nil
 }
 
-func sortAlgorithm(s string) string {
+func (c *cmd) sortAlgorithm(s string) string {
 	if len(s) == 0 {
 		return "" // edge case mimics coreutils
 	}
-	if s[len(s)-1] == '\n' {
-		s = s[:len(s)-1] // remove newline terminator
-	}
-	lines := strings.Split(string(s), "\n")
-	if *reverse {
+	lines := strings.Split(s, "\n")
+	if c.params.reverse {
 		sort.Sort(sort.Reverse(sort.StringSlice(lines)))
 	} else {
 		sort.Strings(lines)
 	}
+
+	if c.params.unique && len(lines) > 1 {
+		j := 1
+		for i := 1; i < len(lines); i++ {
+			if lines[i] == lines[i-1] {
+				continue
+			}
+			lines[j] = lines[i]
+			j++
+		}
+
+		lines = lines[:j]
+	}
+
 	return strings.Join(lines, "\n") + "\n" // append newline terminator
 }
 
-func writeOutput(w io.Writer, s string) error {
+func (c *cmd) writeOutput(w io.Writer, s string) error {
 	to := w
-	if *outputFile != "" {
-		if f, err := os.Create(*outputFile); err == nil {
-			to = f
-			defer f.Close()
-		} else {
+	if c.params.outputFile != "" {
+		f, err := os.Create(c.params.outputFile)
+		if err != nil {
 			return err
 		}
+		defer f.Close()
+		to = f
 	}
-	to.Write([]byte(s))
-	return nil
+
+	_, err := to.Write([]byte(s))
+	return err
 }
 
 func main() {
 	flag.Parse()
-	if err := readInput(os.Stdout, os.Stdin, flag.Args()...); err != nil {
+	p := params{reverse: *reverse, ordered: *ordered, outputFile: *outputFile, unique: *unique}
+	if err := command(os.Stdin, os.Stdout, os.Stderr, p, flag.Args()).run(); err != nil {
+		if err == errNotOrdered {
```

Comment:
should you use errors. here? 

Also, does log.Fatal make sense instead of os.Exit? I don't know, so I'm asking :-)

Reply:
log.Fatal will print additional stuff that we don't need there. From POSIX spec: `Same as -c, except that a warning message shall not be sent to standard error if disorder or, with -u, a duplicate key is detected.`

errors.Is will work, but I think we explicitly search for specific error here, so `==` should be ok.